utils/lib.sh

```diff
@@ -496,6 +502,178 @@ service_is_available() {
+PYDIST="${PYDIST:=dist}"
+# folder where the intermediate build files take place
+PYBUILD="${PYBUILD:=build/py${PY}}"
+
+# list of python packages (folders) or modules (files) installed by command:
+# pyenv.install
+PYOBJECTS="${PYOBJECTS:=.}"
+
+# https://www.python.org/dev/peps/pep-0508/#extras
+#PY_SETUP_EXTRAS='[develop,test]'
+PY_SETUP_EXTRAS="${PY_SETUP_EXTRAS:=[develop,test]}"
+
+pyenv() {
+
+    # usage:  pyenv [vtenv_opts ...]
+    #
+    #   vtenv_opts: see 'pip install --help'
+    #
+    #   PYENV     [virtualenv] usage: source ./local/py3/bin/activate
+    #   PYENV     [virtualenv] installing requirements*.txt into ./local/py3
+    #   ...
+    #   Successfully installed ...
+    #
+    # Build virtualenv with 'requirements*.txt' (PY_ENV_REQ) installed.  The
+    # virtualenv will be reused by validating sha256sum of the requirement
+    # files:
+    #
+    #   ...
+    #   PYENV     [virtualenv] usage: source ./local/py3/bin/activate
+    #   PYENV     [virtualenv] using ./local/py3 // glob pattern requirements*.txt --> sha256 OK
+    #             [virtualenv] - 6767d1ae72ee763f206eed702513b87f4c40c9bc5d4fc5d0c01961b355299ac9  requirements-dev.txt
+    #             [virtualenv] - 329569cde3fd74dfb109fd8da0d5d068747496feb4dae85a7701335cccd8a43a  requirements.txt
+
+    local PYTHON="${PYTHON:=python$PY}"
+
+    required_commands \
+        sha256sum "${PYTHON}" \
+        || exit
+
+    local PY_ENV_REQ="${PY_ENV_REQ:=${REPO_ROOT}/requirements*.txt}"
+    local VTENV_OPTS="$*"
+    local pip_req=""
+    local sum_req=""
+
+    if [ ! -f "${PY_ENV}/requirements.sha256" ] \
+        || ! sha256sum --check --status <"${PY_ENV}/requirements.sha256" 2>/dev/null; then
+
+        build_msg PYENV "[virtualenv] installing ${PY_ENV_REQ} into ${PY_ENV}"
+        rm -f "${PY_ENV}/${PY_ENV_REQ}.sha256"
+
+        # shellcheck disable=SC2086
+        for i in ${PY_ENV_REQ}; do pip_req="$pip_req -r $i"; done
+        # shellcheck disable=SC2086
+        for i in ${PY_ENV_REQ}; do sum_req="$sum_req $i"; done
+        # shellcheck disable=SC2086
+        "${PYTHON}" -m venv ${VTENV_OPTS} "${PY_ENV}"
+        "${PY_ENV_BIN}/python" -m pip install -U pip wheel setuptools
+        # shellcheck disable=SC2086
+        "${PY_ENV_BIN}/python" -m pip install $pip_req \
+            && sha256sum $sum_req > "${PY_ENV}/requirements.sha256"
+
+    else
+        build_msg PYENV "[virtualenv] using ${PY_ENV} // glob pattern ${PY_ENV_REQ} --> sha256 OK"
+        sed 's/^/          [virtualenv] - /' <"${PY_ENV}/requirements.sha256"
+    fi
+    build_msg PYENV "[virtualenv] usage: source ${PY_ENV}/bin/activate"
+}
+
+pyenv.drop() {
+
+    # usage:  pyenv.drop
+
+    build_msg PYENV "[virtualenv] drop ${PY_ENV}"
+    rm -rf "${PY_ENV}"
+}
+
+# shellcheck disable=SC2120
+pyenv.install() {
+
+    # usage:  pyenv.install
+
+    build_msg PYENV "[pyenv.install] install packages: ${PYOBJECTS}"
+    pyenv "$@"
+
+    local py_imp=""
```

Comment:
Now we have a less generic one in the manage script https://github.com/return42/searx/blob/manage-script/manage#L342-L357 

Since `pyenv.install` is a elementary prerequisite the script will die if the installation fails in any way (installation can not be assured).